R/calc.dtsurvey.R

```diff
@@ -0,0 +1,364 @@
+         med_fun = med_fun,
+         total_fun = total_fun,
+         num_fun = num_fun,
+         denom_fun = denom_fun,
+         mis_fun = mis_fun,
+         misp_fun = misp_fun,
+         ut_fun = ut_fun,
+         obs_fun = obs_fun))
+
+  #remove nulls
+  the_call = as.list(the_call)
+  nulls = vapply(the_call, is.null, TRUE)
+  the_call = the_call[!nulls]
+
+  #Remove numerator if x is a factor since that gets calculated differently
+  if(xisfactor) the_call = the_call[which(!names(the_call) %in% 'numerator')]
+  the_call = as.call(the_call)
+
+  #compute the aggregations
+  if(!xisfactor){
+    res = DT[, ccc,
+      by = by,
+      env = list(ccc = the_call)][, level:=NA]
+  }else{
+    r1 = DT[, ccc,
+    by = by,
+    env = list(ccc = the_call)]
+
+    #for factors, the numerator needs to be calculated separately per level.
+    r2 = DT[, list(
+      numerator = .N
+    ),
+    by = c(by, x)]
+    setnames(r2, x, 'level')
+
+    r1[, id := .I]
+
+  }
+
+  #if mean/total were asked for, split out things
+  if(any(c('rate', 'mean') %in% metrics)){
+
+    if(xisfactor){
+      r1m = r1[, unlist(mean, recursive = FALSE), id]
+      setnames(r1m, c('id', 'mean', 'mean_se', 'mean_lower', 'mean_upper', 'level'))
+
+    }else{
+      r1m = NULL
+      res[, c('mean', 'mean_se', 'mean_lower', 'mean_upper') := rbindlist(mean)]
+    }
+  }
+
+  if('total' %in% metrics){
+
+    if(xisfactor){
+      r1t = r1[, unlist(total, recursive = FALSE), id]
+      setnames(r1t, c('id', 'total', 'total_se', 'total_lower', 'total_upper', 'level'))
+
+    }else{
+      r1t = NULL
+      res[, c('total', 'total_se', 'total_lower', 'total_upper') := rbindlist(total)]
+    }
+
+  }
+
+  #assemble factors
+  if(any(c('mean', 'rate', 'total') %in% metrics) && xisfactor){
+    if(exists('r1m')){
+      r1 = merge(r1[, mean:=NULL], r1m, by = 'id')
+    }
+    if(exists('r1t')){
+      byby = c('id', 'level')
+      byby = byby[c(T, 'level' %in% names(r1))]
+      r1 = merge(r1[, total:= NULL], r1t, by= byby)
+    }
+  }
+  if(xisfactor){
+    if(!any(c('rate','total','mean') %in% metrics)){
+      r1[, one := 1]
+      r2[, one := 1]
+
+      res = merge(r1, r2, by = c('one',by), all.x = T)
+      res[, one := NULL]
+    }else if('numerator' %in% metrics){
+      res = merge(r1,r2, by = c(by, 'level'), all.x = T)
+    } else{
+      res = r1
+    }
+    res[, id := NULL]
+  }
+
+  #if asked for, compute rse and rate
+  if('rse' %in% metrics){
+    res[, rse := 100*(mean_se / mean)]
+  }
+
+  if('rate' %in% metrics){
+    res[, c('rate', paste0('rate', c('_se', '_lower', '_upper'))) := .SD * per,
+        .SDcols = c('mean', paste0('mean_',c('se', 'lower', 'upper')))]
+    res[, rate_per := per]
+
+    if(!'mean' %in% metrics) res[, c('mean', paste0('mean_',c('se', 'lower', 'upper'))) := NULL]
+
+  }
+
+
+  if(!is.null(time_var)) data.table::setnames(res, 'time', time_var)
+
+  return(res)
+
+
+}
+
+
+# b = data.table(1:10, byvar = 1:2)
```

Comment:
these examples aren't functional

Reply:
Can we delete lines 376-378? I know they won't show up, for the end user, but they don't work (line 377 needs a `what` and line 378 refers to an object that doesn't exist).